pkg/auth/helper.go

```diff
@@ -0,0 +1,213 @@
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+
+	"github.com/gin-gonic/gin"
+	"golang.org/x/crypto/bcrypt"
+	"k8s.io/klog/v2"
+
+	"tkestack.io/kstone/pkg/controllers/util"
+)
+
+const (
+	JWTTokenKey = "kstone-api-jwt"
+
+	signMethodRS256 = "RS256"
+
+	DataSuccess            = "success"
+	DataUnauthorized       = "failed to authenticate user"
+	DataInternalSeverError = "internal server error"
+	UserUnknown            = "unknown"
+
+	PrivateKey = "private.key"
+
+	DefaultConfigMapName   = "kstone-api-user"
+	DefaultSecretName      = "kstone-api-key"
+	DefaultKstoneNamespace = "kstone"
+	DefaultUsername        = "admin"
+	DefaultPassword        = "adm1n@kstone.io"
+)
+
+func successResponse(username string, data string) *Response {
+	return &Response{
+		Username: username,
+		Data:     data,
+	}
+}
+
+func successResetPasswordResponse(username string, data string) *Response {
+	return &Response{
+		Username:      username,
+		ResetPassword: true,
+		Data:          data,
+	}
+}
+
+func unauthenticatedResponse() *Response {
+	return &Response{
+		Username: UserUnknown,
+		Data:     DataUnauthorized,
+	}
+}
+
+func internalServerErrorResponse(username string) *Response {
+	return &Response{
+		Username: username,
+		Data:     DataInternalSeverError,
+	}
+}
+
+func LoginRequest(ctx *gin.Context) (*Response, bool, error) {
+	username, password, err := getUser(ctx)
+	if err != nil {
+		klog.Errorf("get user error: %v", err)
+		return unauthenticatedResponse(), false, errors.New(DataUnauthorized)
+	}
+
+	authGetter := initDefaultAuthenticatorInfoGetter(util.NewSimpleClientBuilder(""))
+	a, err := authGetter.Config(DefaultKstoneNamespace, DefaultConfigMapName, DefaultSecretName, username)
+	if err != nil {
+		klog.Errorf("get authenticator error: %v", err)
+		return unauthenticatedResponse(), false, errors.New(DataUnauthorized)
+	}
+
+	if err := checkPassword(a.Password, password); err != nil {
+		klog.Errorf("check password error: %v", err)
+		return unauthenticatedResponse(), false, errors.New(DataUnauthorized)
+	}
+
+	if username != a.Username {
+		return unauthenticatedResponse(), false, errors.New(DataUnauthorized)
+	}
+
+	a.IsLogin = true
+	return a.AuthenticateRequest(ctx)
+}
+
+func LogoutRequest(ctx *gin.Context) (*Response, error) {
+	return successResponse("", DataSuccess), nil
+}
+
+func MiddlewareRequest(ctx *gin.Context) (*Response, bool, error) {
+	authGetter := initDefaultAuthenticatorInfoGetter(util.NewSimpleClientBuilder(""))
+	a, err := authGetter.Config(DefaultKstoneNamespace, DefaultConfigMapName, DefaultSecretName, "")
+	if err != nil {
+		klog.Errorf("get authenticator error: %v", err)
+		return unauthenticatedResponse(), false, err
+	}
+	j, err := initJWTTokenAuthenticator(
```

Comment:
The overall implementation needs to be more general and configurable， we will support more authenticate plugins, refer to https://github.com/tkestack/kstone/blob/master/pkg/featureprovider/plugins.go